src/providers/route.js

```diff
@@ -0,0 +1,244 @@
+angular.module('bicker_router').provider('Route', function(ObjectHelper) {
+  "ngInject";
+  let tokens = {};
+  let urlWriters = {};
+  let urls = [];
+  let persistentStates = [];
+  let ready = false;
+  let types = {};
+  let html5Mode = false;
+
+  let provider = {
+
+    registerType(name, config) {
+      types[name] = config;
+      types[name].regex = new RegExp(types[name].regex.source, 'i');
+      return _.extend({ and: this.registerType }, this);
+    },
+
+    registerUrlToken(name, config) {
+      tokens[name] = _.extend({name}, config);
+      return _.extend({ and: this.registerUrlToken }, this);
+    },
+
+    registerUrlWriter(name, fn) {
+      urlWriters[name] = fn;
+      return _.extend({ and: this.registerUrlWriter }, this);
+    },
+
+    registerUrl(pattern, config) {
+      if (config == null) { config = {}; }
+      let urlData = {
+        compiledUrl: this._compileUrlPattern(pattern, config),
+        pattern
+      };
+
+      urls.push(_.extend(urlData, config));
+      return _.extend({ and: this.registerUrl }, this);
+    },
+
+    setPersistentStates(...stateList) {
+      return (() => {
+        let result = [];
+        for (let state of Array.from(stateList)) {
+          let item;
+          if (!Array.from(persistentStates).includes(state)) { item = persistentStates.push(state); }
+          result.push(item);
+        }
+        return result;
+      })();
+    },
+
+    setHtml5Mode(mode) { return html5Mode = mode; },
+
+    _compileUrlPattern(urlPattern, config) {
+      let match;
+      let compiledUrl;
+      urlPattern = this._escapeRegexSpecialCharacters(urlPattern);
+      urlPattern = this._ensureOptionalTrailingSlash(urlPattern);
+
+      let tokenRegex = /\{([A-Za-z\._0-9]+)\}/g;
+      let urlRegex = urlPattern;
+
+      if (!config.partialMatch) {
+        urlRegex = `^${urlRegex}$`;
+      }
+
+      let tokenList = [];
+
+      while ((match = tokenRegex.exec(urlPattern)) !== null) {
+        let token = tokens[match[1]];
+        tokenList.push(token);
+        urlRegex = urlRegex.replace(match[0], `(${types[token.type].regex.source})`);
+      }
+
+      urlRegex.replace('.', '\\.');
+
```

Comment:
compiledUrl is only ever assigned and used in the return value so it's redundant, please kill it and just have this return an object rather than assigning to compiledUrl and returning it